### PR TITLE
DiHydrogen, Hydrogen, and Aluminum CachedCMakePackage

### PR DESCRIPTION
**Pull request by bvanessen**
Switching the DiHydrogen, Hydrogen, and Aluminum packages to use CachedCMakePackage.

Removed deprecated options, variants, and versions that should no longer be supported.

Updated the LBANN package to interface with the new packages.

Added cmake_cache_filepath and fixed the CMAKE_PREFIX_PATH to return a string not a path, since it is really a list of paths, but not actually a path.

### CODE REVIEW

**Comment by haampie** [comment]
Can you also look into why `lbann` started picking up `dihydrogen@develop`? This happens in Gitlab CI.

It's going to be fixed/disallowed by #39611, but I think using `dihydrogen@develop` was unintentional

**Comment by bvanessen** [comment]
@spackbot fix style

**Review by benson31 — APPROVED** [review]
Looks fine to me. (I only reviewed the package recipes as that's my primary interest here.)

**Comment by bvanessen** [comment]
> Can you also look into why `lbann` started picking up `dihydrogen@develop`? This happens in Gitlab CI.
> 
> It's going to be fixed/disallowed by #39611, but I think using `dihydrogen@develop` was unintentional

Probably because of the set of version constraints for different packages.  We have released new versions, so hopefully this is fixed.

**Comment by bvanessen** [comment]
@haampie This PR is ready to go.

**Review by haampie — APPROVED** [review]
I just glanced over it, assuming you know what you're doing...

**Comment by bvanessen** [comment]
@haampie Thank you.

File: lib/spack/spack/build_systems/cached_cmake.py [hunk 1]
```diff
@@ -257,6 +262,15 @@ def initconfig_hardware_entries(self):
...
             )
+            llvm_bin = spec["llvm-amdgpu"].prefix.bin
+            llvm_prefix = spec["llvm-amdgpu"].prefix
+            # Some ROCm systems seem to point to /<path>/rocm-<ver>/ and
+            # others point to /<path>/rocm-<ver>/llvm
+            if os.path.basename(os.path.normpath(llvm_prefix)) != "llvm":
+                llvm_bin = os.path.join(llvm_prefix, "llvm/bin/")
+            entries.append(
+                cmake_cache_filepath("CMAKE_HIP_COMPILER", os.path.join(llvm_bin, "clang++"))
+            )
```
Comment (adrienbernede) [inline lib/spack/spack/build_systems/cached_cmake.py:273]: @bvanessen do you remember the exact explanation for this ?
I have been in the middle of a revamp in #39613 and now in #41375 instead, and I am trying to find the right way of supporting all the cases.
My understanding is that we need this for recent rocm builds, when the spack spec is asking for +rocm without a proper amdgpu compiler.
However, this is co-existing with the "old fashion way", which is to set HIP_CXX_COMPILER to hipcc (see line 263) and I wonder in it should be the case.

**Pull request merged**